Source/WebCore/Modules/cookie-store/CookieChangeEvent.cpp

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright (C) 2023 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "CookieChangeEvent.h"
+
+#include "CookieChangeEventInit.h"
+#include "CookieListItem.h"
+#include <wtf/Vector.h>
+
+namespace WebCore {
+
+WTF_MAKE_ISO_ALLOCATED_IMPL(CookieChangeEvent);
+
+Ref<CookieChangeEvent> CookieChangeEvent::create(const AtomString& type, CookieChangeEventInit&& eventInitDict)
+{
+    return adoptRef(*new CookieChangeEvent(type, WTFMove(eventInitDict)));
+}
+
+CookieChangeEvent::CookieChangeEvent(const AtomString& type, CookieChangeEventInit&& eventInitDict)
+    : Event(type, Event::CanBubble::No, Event::IsCancelable::No)
```

Comment:
This is actually incorrect, you're failing to pass init to the Event() base constructor. `eventInitDict` is what should control bubbles / cancellable via EventInit.

Reply:
See `BlobEvent` as example.

Reply:
I'm not entirely sure I understand the issue. For example, `ApplePayCancelEvent` does what I have done above. Although, it does make sense that the whoever is creating the `CookieChangeEvent` should have control over canBubbles/isCancelable. I can add two fields to `CookieChangeEventInit' like so:

```
Event::CanBubble canBubble;
Event::IsCancelable isCancelable;
```
  
and then in `CookieChangeEvent::create` I can change it to:
```
CookieChangeEvent::CookieChangeEvent(const AtomString& type, CookieChangeEventInit&& eventInitDict)
    : Event(type, eventInitDict.canBubble, eventInitDict.isCancelable)
    , m_changed(WTFMove(eventInitDict.changed))
    , m_deleted(WTFMove(eventInitDict.deleted))
{ }
```

Would this be preferable?

Reply:
You don’t need to add data members, they are already there.. you inherit them from EventInit.

Reply:
Also Apple Pay code is not necessarily a good reference.